File: src/Net/1. First Course/3. Second Hole/RefactoringGolf.Store/OrderItem.cs
namespace RefactoringGolf.Stack
{
    public class OrderItem
    {
        public Product Product { get; private set; }

        public int Quantity { get; private set; }

        /*
         * Order Item Constructor
         */
        public OrderItem(Product product, int quantity)
        {
            Product = product;
            Quantity = quantity;
        }

        public decimal Total()
        {
            decimal discount = 0;
            if (Product.Category == ProductCategory.Accessories)
            {
                discount = this.CalculateAccessoriesDiscount();
            }
            if (Product.Category == ProductCategory.Bikes)
            {
                discount = this.CalculateBikesDiscount();
            }
            if (Product.Category == ProductCategory.Cloathing)
            {
                discount = this.CalculateCloathingDiscount();
            }
            return this.UnitPricePerQuantity() - discount;
        }

        private decimal CalculateAccessoriesDiscount()
        {
            decimal discount = 0;
            if (this.UnitPricePerQuantity() >= 100)
            {
                discount = this.UnitPricePerQuantity() * 10 / 100;
            }
            return discount;
        }

        private decimal CalculateBikesDiscount()
        {
            return this.UnitPricePerQuantity() * 20 / 100;
        }

        private decimal CalculateCloathingDiscount()
        {
            decimal discount = 0;
            if (this.Quantity > 2)
            {
                discount = this.Product.UnitPrice;
            }
            return discount;
        }

        private decimal UnitPricePerQuantity()
        {
            return this.Product.UnitPrice * this.Quantity;
        }
    }
}

File: src/Java/1. First Course/2. Second Hole/src/main/refactoringgolf/store/OrderItem.java
package refactoringgolf.store;

public class OrderItem {

	private Product product;
	private int quantity;

	/*
	 * Order Item Constructor
	 */
	public OrderItem(Product product, int quantity) {
		this.product = product;
		this.quantity = quantity;
	}

	public Product getProduct() {
		return product;
	}

	public int getQuantity() {
		return quantity;
	}

	public float total() {
		float discount = 0;
		if (getProduct().getCategory() == ProductCategory.Accessories) {
			discount = calculateAccessoriesDiscount();
		}
		if (getProduct().getCategory() == ProductCategory.Bikes) {
			discount = calculateBikesDiscount();
		}
		if (getProduct().getCategory() == ProductCategory.Cloathing) {
			discount = calculateCloathingDiscount();
		}
		return itemAmount() - discount;
	}

	private float calculateAccessoriesDiscount() {
		float discount = 0;
		float unitPricePerQuantity = itemAmount();
		if (unitPricePerQuantity >= 100) {
			discount = unitPricePerQuantity * 10 / 100;
		}
		return discount;
	}

	private float calculateBikesDiscount() {
		return itemAmount() * 20 / 100;
	}

	private float calculateCloathingDiscount() {
		float discount = 0;
		if (getQuantity() > 2) {
			discount = getProduct().getUnitPrice();
		}
		return discount;
	}

	private float itemAmount() {
		return getProduct().getUnitPrice() * getQuantity();
	}
}
